contracts/index_executor/src/task.rs

```diff
@@ -272,221 +420,396 @@ impl Task {
         context: &Context,
         _client: &StorageClient,
     ) -> Result<(), &'static str> {
-        // Only in last step the recipient with be set as real recipient on destchain,
-        // or else it will be the worker account under the hood
         let mut nonce_map: Mapping<String, u64> = Mapping::default();
-        for index in start_index as usize..self.steps.len() {
-            let step = &mut self.steps[index];
-            let nonce = match nonce_map.get(&step.chain) {
-                Some(nonce) => nonce,
-                None => {
-                    let chain = context
-                        .registry
-                        .get_chain(step.chain.clone())
-                        .ok_or("MissingChain")?;
-                    let account_info = context.get_account(self.worker).ok_or("WorkerNotFound")?;
-                    let account = match chain.chain_type {
-                        ChainType::Evm => account_info.account20.to_vec(),
-                        ChainType::Sub => account_info.account32.to_vec(),
-                        // ChainType::Unknown => panic!("chain not supported!"),
-                    };
-                    chain.get_nonce(account).map_err(|_| "FetchNonceFailed")?
-                }
-            };
-            step.nonce = Some(nonce);
+
+        // Apply claim nonce if hasn't claimed
+        if self.claim_nonce.is_none() || !self.has_claimed(context)? {
+            let claim_nonce = self.get_nonce(context, &self.source)?;
+            nonce_map.insert(self.source.clone(), &(claim_nonce + 1));
+            self.claim_nonce = Some(claim_nonce);
+        }
+
+        // Apply nonce for each step
+        for index in start_index as usize..self.merged_steps.len() {
+            let nonce: u64 =
+                match nonce_map.get(&self.merged_steps[index].as_single_step().source_chain) {
```

Comment:
Nit: Let's make the key a variable so you don't need to write `merged_steps[index].as_single_step().source_chain` multiple times.

Reply:
Done